prov/gni/test/av.c

```diff
@@ -317,6 +361,422 @@ Test(av_full_table, lookup)
+static void continuous_lookup(void)
+{
+	int i, ret;
+	pthread_t thread;
+	fi_addr_t addresses[MT_ADDR_COUNT];
+	atomic_t done;
+	void *info[2];
+	const int iters = 17;
+
+	ret = pthread_barrier_init(&mtbar, NULL, 2);
+	cr_assert_eq(ret, 0);
+
+	init_av_lock();
+
+	atomic_initialize(&done, 0);
+
+	memset(addresses, 0, MT_ADDR_COUNT*sizeof(fi_addr_t));
+
+	info[0] = (void *) addresses;
+	info[1] = (void *) &done;
+
+	ret = pthread_create(&thread, NULL, lookup_random, info);
+	cr_assert_eq(ret, 0);
+
+	ret = pthread_barrier_wait(&mtbar);
+	cr_assert((ret == PTHREAD_BARRIER_SERIAL_THREAD) || (ret == 0));
+
+	for (i = 0; i < iters; i++) {
+		for (i = 0; i < MT_ADDR_COUNT; i++) {
+			ret = fi_av_insert(av, (void *) &simple_ep_names[i], 1,
+					   &addresses[i], 0, NULL);
+			cr_assert_eq(ret, 1);
+		}
+		for (i = 0; i < MT_ADDR_COUNT; i++) {
+			ret = fi_av_remove(av, &addresses[i], 1, 0);
+			cr_assert_eq(ret, FI_SUCCESS);
+		}
+	}
+
+	atomic_set(&done, 1);
+
+	ret = pthread_join(thread, NULL);
+	cr_assert_eq(ret, 0);
+
+	ret = pthread_barrier_destroy(&mtbar);
+	cr_assert_eq(ret, 0);
+}
+
+Test(av_full_map, mt_lookup)
+{
+	continuous_lookup();
+}
+
+Test(av_full_table, mt_lookup)
+{
+	continuous_lookup();
+}
+
+static const int state_empty = 1;
+static const int state_full = 2;
+static const int state_locked = 3;
+
+static void *continuous_insert(void *data)
+{
+	int i, pos, n, ret;
+	int expected_state;
+	struct gnix_ep_name *ep_names = ((struct gnix_ep_name **) data)[0];
+	fi_addr_t *addresses = ((fi_addr_t **) data)[1];
+	atomic_t *fe = ((atomic_t **) data)[2];
+	int num_insertions = (int) ((uint64_t *) data)[3];
+	int num_addrs = (int) ((uint64_t *) data)[4];
+	atomic_t *done = ((atomic_t **) data)[5];
+
+	ret = pthread_barrier_wait(&mtbar);
+	if ((ret != PTHREAD_BARRIER_SERIAL_THREAD) && (ret != 0)) {
+		pthread_exit((void *) 1UL);
+	}
+
+	i = 0;
+	pos = 0;
+	while ((i < num_insertions) && !atomic_get(done)) {
+		n = (pos++)%num_addrs;
+		expected_state = state_empty;
+		if (atomic_cas_weak(&fe[n], &expected_state, state_locked)) {
+			av_lock();
+			ret = fi_av_insert(av, (void *) &ep_names[n], 1,
+					   &addresses[n], 0, NULL);
+			av_unlock();
+			if (ret != 1) {
+				/* flag shutdown to avoid deadlock */
+				atomic_set(done, 1);
+				pthread_exit((void *) 1UL);
+			}
+			atomic_set(&fe[n], state_full);
+			i++;
+		}
+	}
+
+	pthread_exit((void *) NULL);
+}
+
+static void *continuous_remove(void *data)
+{
+	int pos, n, ret;
+	int expected_state;
+	fi_addr_t *addresses = ((fi_addr_t **) data)[0];
+	atomic_t *fe = ((atomic_t **) data)[1];
+	int num_addrs = (int) ((uint64_t *) data)[2];
+	atomic_t *done = ((atomic_t **) data)[3];
+
+	ret = pthread_barrier_wait(&mtbar);
+	if ((ret != PTHREAD_BARRIER_SERIAL_THREAD) && (ret != 0)) {
+		pthread_exit((void *) 1UL);
+	}
+
+	pos = 0;
+	while (!atomic_get(done)) {
+		n = (pos++)%num_addrs;
+		expected_state = state_full;
+		if (atomic_cas_weak(&fe[n], &expected_state, state_locked)) {
+			av_lock();
```

Comment:
I was under the impression that the AV tables should be threadsafe. Why are we locking outside of the function calls? Is this an intermediate step in the process to getting to a threadsafe model?